issue 16 - added ci for ubuntu

dapsavoie opened the pull request:
I created  few scripts in ci/ubuntu that tests if the project builds on ubuntu. setup.env sets up profotbuf and rust. I created a build script that calls the setup script and then builds the binaries for civkit-cli, civkitd, and civkit-sample. It captures the build statuses and prints them out. If any build fails, it exits with an appropriate message.

ariard commented:
Looks strong, just note you have the `Cargo.lock` committed in the branch.

Will test on Ubuntu if the script works.

dapsavoie commented:
Removed cargo.lock. let me know how it goes!

ariard commented:
Sorry still late on review here, despite saying I’ll review today. Writing my first NIP took me longer than expected :)

ariard commented on ci/ubuntu/setup_env.sh:
```diff
@@ -0,0 +1,12 @@
+sudo apt-get install -y protobuf-compiler
+
+# Install rustup, the Rust toolchain installer
+curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
```
we should commit to a rust MSRV in sync with our dependencies, currently the one supported by LDK is 1.41.1 and then a recent compiler version like 1.68.0 or 1.69.0 for dev/testing purposes, though discard the latest one served by rustup iirc ?

dapsavoie replied:
I made a commit with these lines
```
# Install the minimum supported Rust version (MSRV)
rustup install 1.41.1

# Install a newer version of Rust for development/testing
rustup install 1.68.0
```
hopefully that resolves this ?

ariard replied:
yes all good that way we’ll have a stable MSRV and current rust version is 1.70.0 so we’ll be able to stick to 1.68.0 for a while

ariard commented on ci/ubuntu/run.sh:
```diff
@@ -0,0 +1,39 @@
+echo "Building civkit-sample..."
+SAMPLE_STATUS=$(./build_civkit-sample.sh)
+build_statuses+=("$SAMPLE_STATUS")
+
```
This sounds all good to me, we have at the top `run.sh`:
- calling `setup_env.sh` to configure the good rust env
- calling `build_civkit-cli.sh` for `civkit-cli`
- calling `build_civkitd.sh` for `civkitd`
- calling `build_civkit-sample.sh` for `civkit-sample`

I think this is still missing a `.github/workflows/build.yml` to do the linkage with Github Actions.

dapsavoie replied:
This code is for running such tests in a local setup. 

I have committed a change to run the binary tests for the resources civkitd, cli and sample) on ubuntu in this branch. the setup for a local run test vs github actions is slightly different. Rather then make life complicated, I propose we have the option for local runs with GA actions configured to automate this. 

Feel free to review here the output of a GA test here
https://github.com/dapsavoie/civkit-node/actions/runs/5475581696

ariard reviewed (commented):
Looking into the Github Actions documentation right now to see how it works and what the settings I need to kick, they have a bunch of options for open-source projects, though we can cover the cycles.

Once I understand how Github Actions works and we’re good on the rust compiler version, can land it.

ramin commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,38 @@
+
+    strategy:
+      matrix:
+        os: [ubuntu-latest]
```
it would be better to bind this to a specific version than let it drift and change with "latest" which might break unexpectedly as "latest" moves forward. I'd suggest explicitly binding to the current version only eg: `ubuntu-22.04`

dapsavoie replied:
Done

ramin commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,38 @@
+      with:
+        toolchain: ${{ matrix.rust }}
+        override: true
+
```
I'd recommend adding a cache step here, simply:

```
- uses: Swatinem/rust-cache@v2
```

will speed up subsequent builds of this pipeline SIGNIFICANTLY

dapsavoie replied:
I have added that as well! Thanks for the feedback. Let me know and if it looks good to you. The actions ran successfully.

dapsavoie replied:
That change reduced build speed from 2m48s to 1m! amazing!

dapsavoie commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,38 @@
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Rust
+      uses: actions-rs/toolchain@v1
```
Great suggestion, thanks. Ive made that change in the ci.yml.

ariard commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,73 @@
+
+  build_debian:
+    # This job tests on Debian.
+    runs-on: ubuntu-22.04
```
if you can add the comment as you’re saying on the thread “debian is run in a docker on ubuntu as it’s not natively supported by GA runners”

ariard commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,73 @@
+    runs-on: ubuntu-22.04
+
+    container:
+      image: debian:latest
```
if we can pin explicitly the debian version like bullseye otherwise we’ll have serve the GA runners definition of “latest”. most of debian platforms are probably on bullseye right now

dapsavoie replied:
ive added that as a comment on the code.

ariard commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,73 @@
+
+    - name: Install Rust Toolchain (stable)
+      run: |
+          rustup toolchain install stable --profile minimal
```
well i think here the rustup toolchain should switch to both the MSRV and 1.68.0, happy to take this change in a follow-up PR

dapsavoie replied:
I did briefly test this but I agree, lets work on this in a seperate PR. I tried to implement a quicker solution but since this is quite well developed, its best to finish this properly then address that rust version setup. Would you want me to open a new issue?

ariard replied:
Yes feel free to open a new issue or open pr as you wish.

ariard reviewed (commented):
Tested the scripts and read the documentation on GA here: https://docs.github.com/en/actions/quickstart

Sounds good to me, added few comments though can be added as a follow-up.

Feel free to squash the 36 commits in a few ones before I can merge it.

dapsavoie commented:
> Tested the scripts and read the documentation on GA here: https://docs.github.com/en/actions/quickstart
> 
> Sounds good to me, added few comments though can be added as a follow-up.
> 
> Feel free to squash the 36 commits in a few ones then I'll merge it.

I squashed the commits. It was a bit of testing going on to figure out the working format! I noticed debian is failing due to tokio-socks crate. I believe this is an expected behaviour and indicates the issue this CI should be indiciating.

I will let you review and hopefully its merge-able now. Happy to address further issue on rust versions and create the issue.

Let me know what you think about merging and any other comments. We can address the rust versions seperately to keep it. 

I am grateful for the opportunity to contribute.

ariard reviewed (approved):
ACK f293d37

Let’s see what can be improved with future PRs or bookmark them in issues, having things working on ubuntu is already a nice step forward.

ariard commented:
> I am grateful for the opportunity to contribute.

This is me who is thankful for the nice contribution here.
